src/commcare_cloud/ansible/plugins/lookup/cchq_aws_secret.py

```diff
@@ -0,0 +1,91 @@
+
+
+class LookupModule(aws_secret.LookupModule):
+    """
+    commcare-cloud's wrapper around aws_secret
+
+    Adds commcare-cloud specific caching to aws_secret lookup plugin for big speedup
+    Also assumes the value is json and returns the string json-decoded unlike aws_secret.
+
+    To cache even between forks of the same process/run, we use a one-time process-wide key
+    to encrypt values and write them to disk in files <env_dir>/.generated/.
+    After the run is over, the encryption key is lost to time and the files are useless
+    and indecipherable even by the person who ran it.
+    """
+    def run(self, terms, variables, **kwargs):
+        terms = tuple(terms)
+        term, = terms
+        value = self.get_cache(term, inventory_dir=variables['inventory_dir'])
+        try:
+            if value is Ellipsis:
+                logging.debug('Fetching secret {}'.format(terms))
+                try:
+                    value = super(LookupModule, self).run(terms, variables, **kwargs)
+                except Exception as e:
+                    logging.debug('Caching error fetching secret: {}'.format(e))
+                    value = e
+                self.set_cache(term, value, inventory_dir=variables['inventory_dir'])
+        except Exception as e:
+            logging.warn('There was an error fetching or caching the secret: {}'.format(e))
+            raise
+
+        if isinstance(value, Exception):
+            raise value
+        else:
+            return [json.loads(item) for item in value]
+
+    def get_cache(self, term, inventory_dir):
+        while True:
+            # - If this fork is the first one to do the lookup, it'll meet an empty cache
+            # and claim it by writing the special value LOOKUP_IN_PROGRESS_ON_ANOTHER_FORK to it,
+            # and then write the result to the cache when it's done.
+            # - If this fork is not the first one to do the lookup, it'll either
+            #   - see LOOKUP_IN_PROGRESS_ON_ANOTHER_FORK in the cache and wait to try again and then eventually
+            #   - see a value in the cache and use that value
+            try:
+                with open(self._secrets_cache_filename(term, inventory_dir=inventory_dir), 'rb') as f:
+                    contents = f.read()
+                    if contents == LOOKUP_IN_PROGRESS_ON_ANOTHER_FORK:
+                        # If another fork is in the progress of looking this up
+                        # wait a short while and then try again from the top of the while True loop
+                        time.sleep(.100)
+                        continue
+                    return cPickle.loads(Fernet(self._encryption_key()).decrypt(contents))
+            except (IOError, InvalidToken, cPickle.UnpicklingError):
+                with open(self._secrets_cache_filename(term, inventory_dir=inventory_dir), 'wb') as f:
+                    f.write(LOOKUP_IN_PROGRESS_ON_ANOTHER_FORK)
```

Comment:
There's a race condition here where two processes both try to read the cache, but it's empty, so they both start writing. Is there any chance that could cause a bad outcome?

Reply:
In the case that two processes wrote to the same file at the same time, leaving it in a corrupted state, other threads would try to read it, fail to decrypt and unpickle the corrupted binary, and just opt to fetch it fresh themselves. That's obviously less ideal than the caching working, but my estimation is that it's fairly low and when it does happen there's no real catastrophic effect.